Lib/test/test_shelve.py

```diff
@@ -165,6 +168,298 @@ def test_default_protocol(self):
+                with shelve.open(
+                    self.fn,
+                    protocol=proto,
+                    serializer=serializer,
+                    deserializer=deserializer,
+                ) as s:
+                    num = 1
+                    bytes_data = b"Hello, world!"
+                    bytearray_data = bytearray(b"\x00\x01\x02\x03\x04")
+                    array_data = array.array("i", [1, 2, 3, 4, 5])
+                    memoryview_data = memoryview(b"abcdefgh")
+
+                    s["number"] = num
+                    s["bytes_data"] = bytes_data
+                    s["bytearray_data"] = bytearray_data
+                    s["array_data"] = array_data
+                    s["memoryview_data"] = memoryview_data
+
+                    if proto == 5:
+                        self.assertEqual(s["number"], str(num))
+                        self.assertEqual(s["bytes_data"], "Hello, world!")
+                        self.assertEqual(
+                            s["bytearray_data"], bytearray_data.decode()
+                        )
+                        self.assertEqual(
+                            s["array_data"], array_data.tobytes().decode()
+                        )
+                        self.assertEqual(s["memoryview_data"], "abcdefgh")
+                    else:
+                        self.assertEqual(s["number"], "int")
+                        self.assertEqual(s["bytes_data"], "bytes")
+                        self.assertEqual(s["bytearray_data"], "bytearray")
+                        self.assertEqual(
+                            s["array_data"], array_data.tobytes().decode()
+                        )
+                        self.assertEqual(s["memoryview_data"], "memoryview")
+
+        with self.assertRaises(AssertionError):
+            def serializer(obj, protocol=None):
+                return bytes(f"{type(obj).__name__}", 'utf-8')
+
+            def deserializer(data):
+                pass
+
+            with shelve.open(self.fn,
+                             serializer=serializer,
+                             deserializer=deserializer) as s:
+                s['number'] = 100
+                self.assertEqual(s['number'], 100)
+
+    def test_custom_incomplete_serializer_and_deserializer(self):
+        dbm_sqlite3 = import_helper.import_module("dbm.sqlite3")
+        os.mkdir(self.dirname)
+        self.addCleanup(os_helper.rmtree, self.dirname)
+
+        with self.assertRaises(dbm_sqlite3.error):
+            def serializer(obj, protocol=None):
+                pass
+
+            def deserializer(data):
+                return BytesIO(data).read().decode("utf-8")
+
+            with shelve.open(self.fn,
+                             serializer=serializer,
+                             deserializer=deserializer) as s:
+                s['number'] = 100
+                self.assertEqual(s['number'], 100)
+
+    def test_custom_serializer_and_deserializer_bsd_db_shelf(self):
```

Comment:
Should be only available if `berkeleydb` exists.

Reply:
If the `import_module` function cannot find the `berkeleydb`, it already skips the test.

Reply:
Oh ok I wasn't aware of! Thanks.